core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,325 @@
+
+    public DuplicateByteSequenceSpotter() {
+        this.nodesAllocatedByDepth = new int[4];
+        this.bytesAllocated = 0;
+        root = new RootTreeNode((byte) 1, null, 0);
+    }
+
+    /**
+     * Reset the sequence detection logic to avoid any continuation of the
+     * immediately previous bytes. A minimum of dupSequenceSize bytes need to be
+     * added before any new duplicate sequences will be reported.
+     */
+    public void startNewSequence() {
+        sequenceBufferFilled = false;
+        nextFreePos = 0;
+    }
+
+    /**
+     * 
+     * @param b
+     *            the next byte in a sequence
+     * @return number of times this byte and the preceding 6 bytes have been
+     *         seen before as a sequence
+     * 
+     */
+    public short addByte(byte b) {
+        // Add latest byte to circular buffer
+        sequenceBuffer[nextFreePos] = b;
+        nextFreePos++;
+        if (nextFreePos >= sequenceBuffer.length) {
+            nextFreePos = 0;
+            sequenceBufferFilled = true;
+        }
+        if (!sequenceBufferFilled) {
+            return 0;
+        }
+        TreeNode node = root;
+        // replay updated sequence of bytes represented in the circular
+        // buffer starting from the tail
+        int p = nextFreePos;
+
+        // The first tier of nodes are addressed using individual bytes from the
+        // sequence
+        node = node.add(sequenceBuffer[p], 0);
+        p = nextBufferPos(p);
+        node = node.add(sequenceBuffer[p], 1);
+        p = nextBufferPos(p);
+        node = node.add(sequenceBuffer[p], 2);
+
+        // The final 3 bytes in the sequence are represented by a single node.
+
+        // Fill last 3 bytes in search buffer
+        for (int i = 0; i < 3; i++) {
+            p = nextBufferPos(p);
+            lastSequenceSearchBuffer[i] = sequenceBuffer[p];
+        }
+        // The last byte is unused
+        assert (lastSequenceSearchBuffer[3] == 0);
+        short hitCount = node.add(byteSequenceAsInt(lastSequenceSearchBuffer));
+        return (short) (hitCount - 1);
+    }
+
+    // packing an array of 4 bytes to an int, big endian
+    private static int byteSequenceAsInt(byte[] bytes) {
+        return bytes[0] << 24 | (bytes[1] & 0xFF) << 16 | (bytes[2] & 0xFF) << 8 | (bytes[3] & 0xFF);
+    }
+
+    private int nextBufferPos(int p) {
+        p++;
+        if (p >= sequenceBuffer.length) {
+            p = 0;
+        }
+        return p;
+    }
+
+    // Base class for nodes in the tree. Subclasses are optimised for use
+    // at different locations in the tree - speed-optimized nodes represent
+    // branches near the root while space-optimized nodes are used for
+    // deeper leaves/branches.
+    abstract class TreeNode {
+
+        TreeNode(byte key, TreeNode parentNode, int depth) {
+            nodesAllocatedByDepth[depth]++;
+        }
+
+        public abstract TreeNode add(byte b, int depth);
```

Comment:
Agreed. I think this class hierarchy stems from earlier designs when I had some more TreeNode specializations and before I replaced them with a single int for leaf nodes